### PR TITLE
'agent-specified' needs to be an exception for environment names with dashes

### PR DESCRIPTION
**Pull request by WhatsARanjit**
… dashes

### CODE REVIEW

**Comment by WhatsARanjit** [comment]
@smbambling Taking a stab at this.  Let me know if you have comments:

https://github.com/WhatsARanjit/prosvcs-node_manager/commit/e83365cf31cbf94f687ff446c43938754587d97d

You can do `node_groups()` and receive all of them, or `node_groups('PE Master')` and receive just one.  As for filtering, I'd probably leave it up to all the other functions to mess with the hash it returns.

**Comment by smbambling** [comment]
Great I'll play with this tonight or early i. The AM

File: lib/puppet/type/puppet_environment.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@
   newparam(:name, :namevar => true) do
     desc 'This is the name of the environment'
     validate do |value|
-      fail("#{value} is not a valid group name") unless value =~ /\A[a-z0-9_]+\Z/
+      fail("#{value} is not a valid group name") unless value =~ /\A[a-z0-9_]+\Z/ or value == 'agent-specified'
```
Comment (reidmv) [inline lib/puppet/type/puppet_environment.rb]: "not a valid **environment** name", yes?

Comment (reidmv) [inline lib/puppet/type/puppet_environment.rb]: er, this looks deeper than I thought. Might be I'm completely off here. Let me know if I am, or I'll figure it out in a minute after some spelunking. :-)

Comment (WhatsARanjit) [inline lib/puppet/type/puppet_environment.rb]: Update the error message to say 'environment name'

**Comment by smbambling** [comment]
@WhatsARanjit: the node_groups() function works pretty well.  With some help from @Volcane in IRC, I'm able to return an array of Environment Groups from the classifier

```puppet
$pe_environment_groups = delete(node_groups().filter |$key, $val| { $val["environment_trumps"] == true }.map |$key, $val| { $val["name"] }, 'Agent-specified environment')
```

**Comment by smbambling** [comment]
Here is a gist of the class that now auto creates and removes classifier env groups based on the availability of a PuppetServer environment

https://gist.github.com/smbambling/29bac41c434274afbfa1

**Pull request merged**